### ISSUE TITLE
[Lab 7 Level 4] Return type of flatMap()

### ISSUE THREAD

**Issue by Xatel** [nus-cs2030/2021-s2#152]
# Summary

So flatmap takes in a function of the form Function<? super T, ? extends Logger<? extends U>>. What I am stuck at is what is the scope of resulting output. Not sure if it is correct but my Logger interface has a flatmap that returns a Logger<? extends U> and my LoggerImpl returns a LoggerImpl<? extends U>. It works when I create a method chain but it gives me this error

incompatible types: no instance(s) of type variable(s) U,U exist so that Logger<? extends U> conforms to Logger<java.lang.Integer>

when I declare the variable with the scoping.
 Logger<Integer> a = Logger.make(5).flatMap(x -> Logger.make(x).map(y -> y + 2)).flatMap(y -> Logger.make(y).map(z -> z * 10));
I'm assuming somehow that <? extends U> cannot be converted to Integer or the other way round?
Could the typecast below in my code affect it somehow?
LoggerImpl<? extends U> newLogger =  (LoggerImpl<? extends U>) mapper.apply(this.value);

## Description:

Clarifications or details with respect to the issue you are facing

## Code Snippets

Attach code snippets (if any)

## Screenshots

Screenshots (if any)

**Comment by gwajoon** [comment]
Hi Xatel, your Logger interface should return a <U>LoggerImpl<U> instead of a Logger<? extends U>. You are trying to return a Logger when your method called in LoggerImpl returns a LoggerImpl, I think that's why you are getting this incompatible type error.

**Comment by Xatel** [comment]
Hmm thats weird. When i tried changing the return type in the Logger interface to just LoggerImpl<? extends U>, both files could compile. 
But when i run jshell Logger.java LoggerImpl.java, it gives an error that it can't find symbol of Logger.
Before changing the code, I also tried LoggerImpl a = Logger.make(5).flatMap(x -> Logger.make(x).map(y -> y + 2)).flatMap(y -> Logger.make(y).map(z -> z * 10)). This gives me the same error.
Additionally, I realised that Logger a = Logger.make(5).flatMap(x -> Logger.make(x).map(y -> y + 2)).flatMap(y -> Logger.make(y).map(z -> z * 10)) works. Its the scoping of the declaration Logger<Integer> that gives the previous mentioned error.

**Comment by clementyee303** [comment]
I noticed you are type-casting your mapper.apply(value). I don't think it's needed if your method returns a Logger<U>. Also, I believe you would have needed to do a @SuppressWarnings("unchecked") with your type-cast which I don't think codecrunch allows for this lab submission. 

Answering your error message, I think you have to return the same type as defined in your interface method. Also, I don't think returning generic wildcard types is a good practice.

**Comment by gwajoon** [comment]
Yup as what clementyee303 mentioned, return the same type as defined in your Logger interface to prevent complications.

**Comment by Jason-Ng00** [comment]
Yep I tried to do a @SuppressWarnings("unchecked") somehow for some type-casting at the start and codecrunch won't allow this. And we don't really need to do any type-casting for this lab.
